addons/netfox.extras/network-simulator.gd

```diff
@@ -0,0 +1,188 @@
+extends Node
+
+class_name NetworkSimulator
+
+## Network Simulator 
+## Auto connects launched instances and simulates network conditions like latency and packet loss
+## To use simply add this node to your scene tree and hook up the signals:
+
+## Signal emitted on instance that successfully started a server
+## Initialize your server from this signal
+signal server_created
+
+## Signal emitted on instance that successfully connected as a client
+## Initialize your clients from this signal
+signal client_connected
+
+@export_category("Server")
+## Server listening address. 127.0.0.1 for localhost, use * for all interfaces.
+@export var hostname: String = "127.0.0.1"
+
+## Server port to listen on, udp proxy will use port + 1 if simulating latency or packet loss
+@export var server_port: int = 9999
+
+@export_category("Network Settings")
+## Simulated latency in milliseconds. Total ping time will be double this value (to and from).
+@export_range(0, 200) var latency_ms: int = 0
+## Simulated packet loss percentage
+@export_range(0, 100) var packet_loss_percent: float = 0.0
+
+var enet_peer: ENetMultiplayerPeer = ENetMultiplayerPeer.new()
+
+# UDP proxy
+var proxy_thread: Thread
+var udp_proxy_server: PacketPeerUDP
+var udp_proxy_port: int = 9998
+var rng_packet_loss: RandomNumberGenerator = RandomNumberGenerator.new()
+
+# Connection tracking
+var client_queues: Dictionary = {} # port to data stream  - Array[QueueEntry]
+var client_peers: Dictionary = {} # port to PacketPeerUDP
+var server_to_client_queue: Array[QueueEntry] = []
+
+class QueueEntry:
+	var packet_data: PackedByteArray
+	var queued_at: int
+	var target_port: int
+	
+	func _init(packet: PackedByteArray, timestamp: int, port: int = -1) -> void:
+		self.packet_data = packet
+		self.queued_at = timestamp
+		self.target_port = port
+
+func _ready() -> void:
+	await get_tree().process_frame
+	udp_proxy_port = server_port + 1
+	
+	if OS.has_feature("editor"):
+		var status = try_and_host()
```

Comment:
We could add some logging both to `try_and_host()` and `try_and_join()`. Probably info level to log result, and debug to log it happening and params.